### ISSUE TITLE
Suite / Test instance relationship representation in serialized SciUnit output

### ISSUE THREAD

**Issue by gidili** [MetaCell/geppetto-scidash#33]
Hi @rgerkin as we are working with @NeZanyat on filtering scores based on suite and approaching the building of the suite view, it popped out that we don't have any example of how you would represent the relationship between a suite and a test instance (currently modeled as a many to may) in the serialized output of sciunit / neuronunit. Would this be simply a case of a list of suite-ids on the test instance (multiple foreign key kinda thing), and if so could you update the serialized json in the example to show this?

**Comment by rgerkin** [comment]
Currently the tests know nothing about the suite(s) they may belong to.  One test could belong to one or more (or zero) suites, but there is nothing in the Python object model that identifies this.  Since there is no way to do so in Python, I would be OK with there being no way to do so on the site, but alternatively just sticking with many-to-many is fine as well since that should impose no significant restrictions on the relationships.

**Comment by gidili** [comment]
After our chat, latest on this is that we will drop suite column from score grid view. 

Also we will get a python notebook example from @rgerkin on how to add tests to a suite and an example of said serialised suite.

This will prompt some rework of the upload functionality to upload suite objects pointing to test instances (at the moment it only works with score) and will be the basis for populating the suite grid view.

**Comment by rgerkin** [comment]
Here is the [serialization notebook](https://github.com/scidash/neuronunit/blob/dev/neuronunit/unit_test/serialization_test.ipynb) updated to show how TestSuites are instantiated and serialized.

**Comment by gidili** [comment]
To summarize the latest conversation on this:

- In order to pull together aggregate scores + model information when visualizing suites (suite view, mockup here: https://github.com/MetaCell/geppetto-scidash/issues/14) we need some way to tie suite info to scores. At the moment we have no way of doing this if the user uploads suits that are just pointing to test instances without some kind of matching on test names
- The tool we identified to make this possible is the introduction of ScoreMatrix/ScoreArray entities (at the very least ScoreArray) that ties in suite info with score info and makes it possible to upload everything in one go
- Another thing that we discussed is adding the hash attribute the the serialized test instances to make the test instance entities id-able without recurring to weak name+url matching or similar  
- Example serialization of the above to be provided by Rick so we can use it to introduce the new entity/ies in our domain model and rework the upload functionality in the api to work with ScoreMatrix and/or ScoreArray that will pull in everything when serialized (including score object)
- It will still be possible to upload score objects independently via the api

cc: @NeZanyat @rgerkin

**Comment by rgerkin** [comment]
Implemented in the newest [serialization notebook](https://github.com/scidash/neuronunit/blob/dev/neuronunit/unit_test/serialization_test.ipynb):
- [x] Hashes now stored in JSON (with `add_props=True`)
- [x] JSON dump for a score matrix is now shown

**Comment by gidili** [comment]
Hi @rgerkin as @NeZanyat was working in this he spotted something that may or may not be a feature of the score matrix serialisation - the scores attribute seems to be an array within another array once it's serialised:

<img width="720" alt="screenshot 2018-03-06 09 32 31" src="https://user-images.githubusercontent.com/68134/37024677-7006dcbc-2121-11e8-8b46-f23de8ec0248.png">

It this intended or a glitch of the serialisation library you used for the score matrix?

**Comment by rgerkin** [comment]
The `scores` attribute is 2D array (implemented as a list of lists), so bracket number 1 open the outer list and bracket number 2 opens the inner list.  This is to reflect the idea that `scores` should have shape n_models x n_tests.

**Comment by gidili** [comment]
@rgerkin ah! Got it. Any chance of updating the current example to a 2D score matrix? 😬:smile:

**Comment by rgerkin** [comment]
The current example is 2D; that's why it has 2 brackets.  As for the exact formatting of it, the ScoreMatrix is a subclass of the pandas DataFrame.  When it is jsonified, it uses the custom JSONEncoder [here](https://github.com/scidash/sciunit/blob/dev/sciunit/base.py#L114).  That JSONEncoder handles all SciUnit objects, and those which are subclasses of DataFrames (basically just ScoreMatrices) get handled by lines 122-127.  

Basically the three attributes of the data frame (rows, columns, data) are mapped onto (models, tests, scores).  Since the data is 2D, the scores is 2D.  And the JSONEncoder then just does whatever it does with data in a Pandas dataframe.  I could change that behavior, too, but I'm not sure what that would accomplish.

**Comment by gidili** [comment]
@rgerkin my bad, I see now that we have the 2D score populated in both dimensions 👍 I think we can close this one!

<img width="182" alt="screenshot 2018-03-09 09 19 54" src="https://user-images.githubusercontent.com/68134/37199953-293c61ba-237b-11e8-9682-9ca201455fd3.png">
